## UAFLP Formulations by Konak et al. (2006)

In [1]:
import numpy as np
from docplex.mp.model import Model

### Data Instances:

In [2]:
# Input data: Instance O7
n_dpts = 7
max_bays = np.arange(1, 10)
dpts = np.arange(1, n_dpts + 1)
# width = 13
# height = 8.54
width = 8.54
height = 13
dp_areas = np.array([16, 16, 16, 36, 9, 9, 9])
asp_ratio = np.array([4]*n_dpts)
max_side = np.minimum(np.array([height]*n_dpts), np.sqrt(asp_ratio*dp_areas))
min_side = np.sqrt(dp_areas/asp_ratio)
mat_flows = np.array([[0,  0,  0,  0,  0,  0,  0],
                      [0,  0,  0,  0,  0,  0,  0],
                      [0,  0,  0,  0,  0,  0,  0],
                      [5,  3,  2,  0,  0,  0,  0],
                      [0,  0,  0,  4,  0,  0,  0],
                      [0,  0,  0,  4,  0,  0,  0],
                      [1,  1,  1,  0,  2,  1,  0]])

mat_flows = mat_flows.transpose()

In [ ]:
# Input data: Instance O8
n_dpts = 8
max_bays = np.arange(1, 10)
dpts = np.arange(1, n_dpts + 1)
width = 11.31
height = 13
dp_areas = np.array([16, 16, 16, 36, 36, 9, 9, 9])
asp_ratio = np.array([4]*n_dpts)
max_side = np.minimum(np.array([height]*n_dpts), np.sqrt(asp_ratio*dp_areas))
min_side = np.sqrt(dp_areas/asp_ratio)
mat_flows = np.array([[0,  0,  0,  0,  0,  0,  0,  0],
                      [0,  0,  0,  0,  0,  0,  0,  0],
                      [0,  0,  0,  0,  0,  0,  0,  0],
                      [5,  3,  2,  0,  0,  0,  0,  0],
                      [5,  3,  2,  0,  0,  0,  0,  0],
                      [0,  0,  0,  4,  3,  0,  0,  0],
                      [0,  0,  0,  4,  0,  0,  0,  0],
                      [1,  1,  1,  0,  4,  2,  1,  0]])

mat_flows = mat_flows.transpose()

In [ ]:
# Input data: Instance O9
n_dpts = 9
max_bays = np.arange(1, 10)
dpts = np.arange(1, n_dpts + 1)
width = 12
height = 13
dp_areas = np.array([16, 16, 16, 36, 36, 9, 9, 9, 9])
asp_ratio = np.array([4]*n_dpts)
max_side = np.minimum(np.array([height]*n_dpts), np.sqrt(asp_ratio*dp_areas))
min_side = np.sqrt(dp_areas/asp_ratio)
mat_flows = np.array([[0,  0,  0,  0,  0,  0,  0,  0,  0],
                      [0,  0,  0,  0,  0,  0,  0,  0,  0],
                      [0,  0,  0,  0,  0,  0,  0,  0,  0],
                      [5,  3,  2,  0,  0,  0,  0,  0,  0],
                      [5,  3,  2,  0,  0,  0,  0,  0,  0],
                      [0,  0,  0,  4,  3,  0,  0,  0,  0],
                      [0,  0,  0,  4,  0,  0,  0,  0,  0],
                      [0,  0,  0,  0,  0,  0,  0,  0,  0],
                      [1,  1,  1,  0,  4,  2,  1,  0,  0]])

mat_flows = mat_flows.transpose()

In [2]:
# Input data: Instance vC10R-a
n_dpts = 10
max_bays = np.arange(1, 10)
dpts = np.arange(1, n_dpts + 1)
width = 25
height = 51
dp_areas = np.array([238, 112, 160, 80, 120, 80, 60, 85, 221, 119])
asp_ratio = np.array([5]*n_dpts)
max_side = np.minimum(np.array([height]*n_dpts), np.sqrt(asp_ratio*dp_areas))
min_side = np.sqrt(dp_areas/asp_ratio)
mat_flows = np.array([[  0,   0,  0,   0,   0,  0,  0,   0,  0,  0],
                      [  0,   0,  0,   0,   0,  0,  0,   0,  0,  0],
                      [  0,   0,  0,   0,   0,  0,  0,   0,  0,  0],
                      [  0,   0, 28,   0,   0,  0,  0,   0,  0,  0],
                      [  0,   0, 70,   0,   0,  0,  0,   0,  0,  0],
                      [218, 148,  0,  28,   0,  0,  0,   0,  0,  0],
                      [  0,   0,  0,  70,   0,  0,  0,   0,  0,  0],
                      [  0,   0,  0, 140, 210,  0,  0,   0,  0,  0],
                      [  0, 296,  0,   0,   0,  0,  0,   0,  0,  0],
                      [  0,   0,  0,   0,   0,  0, 28, 888,  59,  0]])

mat_flows = mat_flows.transpose()

In [24]:
# Input data: Instance MB12
n_dpts = 12
max_bays = np.arange(1, 10)
dpts = np.arange(1, n_dpts + 1)
width = 6
height = 8
dp_areas = np.array([1,  1,  1,  1,  1,  1,  1,  1,  4,  4, 16, 16])
asp_ratio = np.array([4]*n_dpts)
max_side = np.minimum(np.array([height]*n_dpts), np.sqrt(asp_ratio*dp_areas))
min_side = np.sqrt(dp_areas/asp_ratio)
mat_flows = np.array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
                      [ 2,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
                      [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
                      [ 0,  5,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
                      [10,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
                      [ 0,  0,  2,  0,  4,  0,  0,  0,  0,  0,  0,  0],
                      [ 0,  0,  9,  0,  0,  0,  0,  0,  0,  0,  0,  0],
                      [ 0,  0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0],
                      [ 9,  0,  0,  0,  0,  4,  0,  0,  0,  0,  0,  0],
                      [ 0,  0,  0,  3,  0,  0,  0,  0,  0,  0,  0,  0],
                      [ 0,  7,  0,  0,  0,  0,  0,  5,  0,  0,  0,  0],
                      [ 0,  2,  0,  3,  5,  0,  0,  0,  3,  0,  1,  0]])

mat_flows = mat_flows.transpose()

In [47]:
# Input data: Instance Garment11
n_dpts = 11
max_bays = np.arange(1, 10)
dpts = np.arange(1, n_dpts + 1)
width = 25
height = 45.60
dp_areas = np.array([50.00, 195.50, 83.27, 100.39, 67.20, 153.00, 153.00, 131.25, 52.90, 52.90, 100.00])
asp_ratio = np.array([4]*n_dpts)
max_side = np.minimum(np.array([height]*n_dpts), np.sqrt(asp_ratio*dp_areas))
min_side = np.sqrt(dp_areas/asp_ratio)
mat_flows = np.array([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
                      [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
                      [  0, 110,   0,   0,   0,   0,   0,   0,   0,   0,   0],
                      [  0, 110,   0,   0,   0,   0,   0,   0,   0,   0,   0],
                      [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
                      [110,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
                      [110,   0,   0, 110,   0,   0,   0,   0,   0,   0,   0],
                      [  0,   0, 110,   0,   0, 110,   0,   0,   0,   0,   0],
                      [ 11,   0,   0,  11,   0,   0,   0,   0,   0,   0,   0],
                      [  0,   0,   0,   0,   4,   0,   0,   0,   4,   0,   0],
                      [  0,   2,   2,   2,   1,   0,   0,   2,   1,   1,   0]])

mat_flows = mat_flows.transpose()

### MIP-FSBP Formulation

In [49]:
# Define UAFLP-MIP model
uaflp_mip = Model('UAFLP-MIP-KONAK-2006')

# Add variables
zs = {(i, k): uaflp_mip.binary_var(name=f'z_{i}_{k}')\
      for i in dpts for k in max_bays}

rs = {(i, j): uaflp_mip.binary_var(name=f'r_{i}_{j}')\
      for i in dpts for j in dpts}

ds = {(k): uaflp_mip.binary_var(name=f'd_{k}')\
      for k in max_bays}

w = {(k): uaflp_mip.continuous_var(name=f'w_{k}', lb=0)\
      for k in max_bays}

ly = {(i): uaflp_mip.continuous_var(name=f'l_y_{i}', lb=0)\
      for i in dpts}

h = {(i, k): uaflp_mip.continuous_var(name=f'h_{i}_{k}', lb=0)\
       for i in dpts for k in max_bays}

x = {(i): uaflp_mip.continuous_var(name=f'x_{i}', lb=0)\
      for i in dpts}

y = {(i): uaflp_mip.continuous_var(name=f'y_{i}', lb=0)\
      for i in dpts}

dx = {(i, j): uaflp_mip.continuous_var(name=f'd_x_{i}_{j}', lb=0)\
        for i in dpts for j in dpts}

dy = {(i, j): uaflp_mip.continuous_var(name=f'd_y_{i}_{j}', lb=0)\
        for i in dpts for j in dpts}

In [50]:
# Add objective function
# total_mf_costs = uaflp_mip.linear_expr()

# for i in dpts:
#     for j in dpts:
#         if i < j:
#             print(i, j, mat_flows[i-1, j-1])
#             total_mf_costs.add_term(dx[i, j], mat_flows[i-1, j-1])
#             total_mf_costs.add_term(dy[i, j], mat_flows[i-1, j-1])


total_mf_costs = uaflp_mip.sum(uaflp_mip.sum(mat_flows[i-1, j-1] * (dx[i, j] + dy[i, j]) for j in dpts if i < j) for i in dpts)

uaflp_mip.minimize(total_mf_costs)

In [51]:
# Add constraints
for i in dpts:
    for j in dpts:
        if i < j:
            uaflp_mip.add_constraint(dx[i, j] >= x[i] - x[j])
            uaflp_mip.add_constraint(dx[i, j] >= x[j] - x[i])
            uaflp_mip.add_constraint(dy[i, j] >= y[i] - y[j])
            uaflp_mip.add_constraint(dy[i, j] >= y[j] - y[i])

In [52]:
for i in dpts:
    uaflp_mip.add_constraint(uaflp_mip.sum(zs[i, k] for k in max_bays) == 1)

In [53]:
for k in max_bays:
    uaflp_mip.add_constraint(w[k] == uaflp_mip.sum(zs[i, k] * dp_areas[i-1] for i in dpts)/height)

In [54]:
for i in dpts:
    for k in max_bays:
        uaflp_mip.add_constraint(min_side[i-1] * zs[i, k] <= w[k])
        uaflp_mip.add_constraint(w[k] <= max_side[i-1] + width * (1 - zs[i, k]))

In [55]:
for i in dpts:
    for k in max_bays:
        uaflp_mip.add_constraint(x[i] >= uaflp_mip.sum(w[j] for j in max_bays if j <= k) - 0.5*w[k] - \
                                 (width - min_side[i-1])*(1 - zs[i, k]))
        
        uaflp_mip.add_constraint(x[i] <= uaflp_mip.sum(w[j] for j in max_bays if j <= k) - 0.5*w[k] + \
                                 (width - min_side[i-1])*(1 - zs[i, k]))

In [56]:
for i in dpts:
    for j in dpts:
        if i < j:
            for k in max_bays:
                uaflp_mip.add_constraint(h[i, k]/dp_areas[i-1] - h[j, k]/dp_areas[j-1] - \
                                         np.max([max_side[i-1]/dp_areas[i-1], max_side[j-1]/dp_areas[j-1]])*(2 - zs[i, k] - zs[j, k]) <= 0)
                uaflp_mip.add_constraint(h[i, k]/dp_areas[i-1] - h[j, k]/dp_areas[j-1] + \
                                         np.max([max_side[i-1]/dp_areas[i-1], max_side[j-1]/dp_areas[j-1]])*(2 - zs[i, k] - zs[j, k]) >= 0)

In [57]:
for k in max_bays:
    uaflp_mip.add_constraint(uaflp_mip.sum(h[i, k] for i in dpts) == height*ds[k])

In [58]:
for i in dpts:
    for k in max_bays:
        uaflp_mip.add_constraint(min_side[i-1]*zs[i, k] <= h[i, k])
        uaflp_mip.add_constraint(h[i, k] <= max_side[i-1]*zs[i, k])

In [59]:
for i in dpts:
    uaflp_mip.add_constraint(uaflp_mip.sum(h[i, k] for k in max_bays) == ly[i])

In [60]:
for i in dpts:
    for j in dpts:
        if i != j:
            uaflp_mip.add_constraint(y[i] - 0.5*ly[i] >= y[j] + 0.5*ly[j] - height*(1 - rs[i, j]))

In [61]:
for i in dpts:
    for j in dpts:
        if i < j:
            uaflp_mip.add_constraint(rs[i, j] + rs[j, i] <= 1)

In [62]:
for i in dpts:
    for j in dpts:
        if i < j:
            for k in max_bays:
                uaflp_mip.add_constraint(rs[i, j] + rs[j, i] >= zs[i, k] + zs[j, k] - 1)

In [63]:
for i in dpts:
    uaflp_mip.add_constraint(0.5*ly[i] <= y[i])
    uaflp_mip.add_constraint(y[i] <= height - 0.5*ly[i])

In [41]:
uaflp_mip.pprint()

// This file has been generated by DOcplex
// model name is: UAFLP-MIP-KONAK-2006
// single vars section
dvar bool z_1_1;
dvar bool z_1_2;
dvar bool z_1_3;
dvar bool z_1_4;
dvar bool z_1_5;
dvar bool z_1_6;
dvar bool z_1_7;
dvar bool z_1_8;
dvar bool z_1_9;
dvar bool z_2_1;
dvar bool z_2_2;
dvar bool z_2_3;
dvar bool z_2_4;
dvar bool z_2_5;
dvar bool z_2_6;
dvar bool z_2_7;
dvar bool z_2_8;
dvar bool z_2_9;
dvar bool z_3_1;
dvar bool z_3_2;
dvar bool z_3_3;
dvar bool z_3_4;
dvar bool z_3_5;
dvar bool z_3_6;
dvar bool z_3_7;
dvar bool z_3_8;
dvar bool z_3_9;
dvar bool z_4_1;
dvar bool z_4_2;
dvar bool z_4_3;
dvar bool z_4_4;
dvar bool z_4_5;
dvar bool z_4_6;
dvar bool z_4_7;
dvar bool z_4_8;
dvar bool z_4_9;
dvar bool z_5_1;
dvar bool z_5_2;
dvar bool z_5_3;
dvar bool z_5_4;
dvar bool z_5_5;
dvar bool z_5_6;
dvar bool z_5_7;
dvar bool z_5_8;
dvar bool z_5_9;
dvar bool z_6_1;
dvar bool z_6_2;
dvar bool z_6_3;
dvar bool z_6_4;
dvar bool z_6_5;
dvar bool z_6_6;
dvar bool z_6_7;
dvar bool z

In [64]:
uaflp_sol = uaflp_mip.solve(log_output=True)

Version identifier: 22.1.0.0 | 2022-03-09 | 1a383f8ce
CPXPARAM_Read_DataCheck                          1
Tried aggregator 1 time.
MIP Presolve eliminated 157 rows and 217 columns.
MIP Presolve modified 3268 coefficients.
Reduced MIP has 2369 rows, 395 columns, and 8455 nonzeros.
Reduced MIP has 218 binaries, 0 generals, 0 SOSs, and 0 indicators.
Presolve time = 0.03 sec. (6.06 ticks)
Probing fixed 0 vars, tightened 4 bounds.
Probing time = 0.02 sec. (3.50 ticks)
Cover probing fixed 0 vars, tightened 35 bounds.
Tried aggregator 1 time.
Detecting symmetries...
MIP Presolve modified 14 coefficients.
Reduced MIP has 2369 rows, 395 columns, and 8455 nonzeros.
Reduced MIP has 218 binaries, 0 generals, 0 SOSs, and 0 indicators.
Presolve time = 0.02 sec. (4.37 ticks)
Probing time = 0.00 sec. (3.24 ticks)
Cover probing fixed 0 vars, tightened 20 bounds.
Clique table members: 165.
MIP emphasis: balance optimality and feasibility.
MIP search method: dynamic search.
Parallel mode: deterministic, u

*     0+    0                        13293.9471        0.0000           100.00%
      0     0        0.0000    89    13293.9471     Cuts: 314      505  100.00%
      0     0        0.0000    32    13293.9471      Cuts: 44      570  100.00%
      0     0        0.0000    51    13293.9471      Cuts: 51      684  100.00%
*     0+    0                        13013.4662        0.0000           100.00%
Detecting symmetries...
      0     2        0.0000    36    13013.4662        0.0000      684  100.00%
Elapsed time = 1.36 sec. (773.33 ticks, tree = 0.02 MB, solutions = 3)
*     5+    1                        13013.4662        0.0000           100.00%
*    13+    5                        13013.4662        0.0000           100.00%
*    64    54      integral     0    12630.1347        0.0000     9915  100.00%
     79    72     9178.1193    35    12630.1347        0.0000    13437  100.00%
    202   148     5507.7918    47    12630.1347        0.0000    25908  100.00%
    355   248        0.00

In [65]:
uaflp_sol.get_objective_value()

8585.08312749297

* O7(v1):

    - Width = 13
    - Height = 8.54
    - OPT: 136.58016035347748

* O7(v2):

    - Width = 8.54
    - Height = 13
    - OPT: 134.19007457625085

* O8:

    - Width = 11.31
    - Height = 13
    - OPT: 245.50564435564434
    - Sol:
    ```
    z_1_9  =  1.0
    z_2_9  =  1.0
    z_3_1  =  1.0
    z_4_9  =  1.0
    z_5_1  =  1.0
    z_6_1  =  1.0
    z_7_9  =  1.0
    z_8_1  =  1.0
    --
    r_1_2  =  1.0
    r_3_5  =  1.0
    r_3_6  =  1.0
    r_3_8  =  1.0
    r_4_1  =  1.0
    r_4_2  =  1.0
    r_6_5  =  1.0
    r_6_8  =  1.0
    r_7_1  =  1.0
    r_7_2  =  1.0
    r_7_4  =  1.0
    r_8_5  =  1.0
    --
    (x, y)  =  (8.346153846153847, 4.0519480519480515)
    (x, y)  =  (8.346153846153847, 1.3506493506493504)
    (x, y)  =  (2.6923076923076925, 11.514285714285714)
    (x, y)  =  (8.346153846153847, 8.441558441558442)
    (x, y)  =  (2.6923076923076925, 3.342857142857143)
    (x, y)  =  (2.6923076923076925, 9.192857142857143)
    (x, y)  =  (8.346153846153847, 12.24025974025974)
    (x, y)  =  (2.6923076923076925, 7.521428571428572)
    ```

* O9:

    - Width = 12
    - Height = 13
    - OPT: 254.3873303167421
    - Sol:
    ```
    z_1_3  =  1.0
    z_2_1  =  1.0
    z_3_9  =  1.0
    z_4_3  =  1.0
    z_5_1  =  1.0
    z_6_1  =  1.0
    z_7_9  =  1.0
    z_8_9  =  1.0
    z_9_1  =  1.0
    --
    r_2_1  =  1.0
    r_2_5  =  1.0
    r_2_6  =  1.0
    r_2_9  =  1.0
    r_4_1  =  1.0
    r_6_5  =  1.0
    r_6_9  =  1.0
    r_7_3  =  1.0
    r_8_3  =  1.0
    r_8_7  =  1.0
    r_9_5  =  1.0
    --
    (x, y)  =  (7.384615384615385, 2.0)
    (x, y)  =  (2.6923076923076925, 11.514285714285714)
    (x, y)  =  (10.692307692307693, 3.058823529411764)
    (x, y)  =  (7.384615384615385, 8.5)
    (x, y)  =  (2.6923076923076925, 3.3428571428571434)
    (x, y)  =  (2.6923076923076925, 9.192857142857143)
    (x, y)  =  (10.692307692307693, 7.838235294117647)
    (x, y)  =  (10.692307692307693, 11.279411764705882)
    (x, y)  =  (2.6923076923076925, 7.521428571428572)
    
    ```
* vC10R-a:

    - Width = 25
    - Height = 51
    - OPT: 20140.35384615384
    - Sol: 
    ```
    z_1_9  =  1.0
    z_2_9  =  1.0
    z_3_7  =  1.0
    z_4_7  =  1.0
    z_5_9  =  1.0
    z_6_9  =  1.0
    z_7_7  =  1.0
    z_8_9  =  1.0
    z_9_9  =  1.0
    z_10_9  =  1.0
    --
    r_1_2  =  1.0
    r_1_5  =  1.0
    r_1_6  =  1.0
    r_1_8  =  1.0
    r_1_9  =  1.0
    r_1_10  =  1.0
    r_2_5  =  1.0
    r_2_8  =  1.0
    r_2_9  =  1.0
    r_2_10  =  1.0
    r_3_4  =  1.0
    r_3_7  =  1.0
    r_6_2  =  1.0
    r_6_5  =  1.0
    r_6_8  =  1.0
    r_6_9  =  1.0
    r_6_10  =  1.0
    r_7_4  =  1.0
    r_8_5  =  1.0
    r_9_5  =  1.0
    r_9_8  =  1.0
    r_9_10  =  1.0
    r_10_5  =  1.0
    r_10_8  =  1.0
    --
    (x, y)  =  (15.441176470588236, 44.77538461538462)
    (x, y)  =  (15.441176470588236, 31.43692307692307)
    (x, y)  =  (2.9411764705882355, 37.4)
    (x, y)  =  (2.9411764705882355, 6.800000000000001)
    (x, y)  =  (15.441176470588236, 3.138461538461538)
    (x, y)  =  (15.441176470588236, 36.45846153846153)
    (x, y)  =  (2.9411764705882355, 18.699999999999996)
    (x, y)  =  (15.441176470588236, 8.499999999999998)
    (x, y)  =  (15.441176470588236, 22.7276923076923)
    (x, y)  =  (15.441176470588236, 13.835384615384612)
    ```

* MB12:

    - Width = 6
    - Height = 8
    - OPT: 124.99999999999993
    - Sol:
    ```
    z_1_6  =  1.0
    z_2_6  =  1.0
    z_3_6  =  1.0
    z_4_6  =  1.0
    z_5_6  =  1.0
    z_6_6  =  1.0
    z_7_6  =  1.0
    z_8_6  =  1.0
    z_9_6  =  1.0
    z_10_6  =  1.0
    z_11_9  =  1.0
    z_12_1  =  1.0
    --
    r_1_9  =  1.0
    r_2_1  =  1.0
    r_2_5  =  1.0
    r_2_6  =  1.0
    r_2_8  =  1.0
    r_2_9  =  1.0
    r_3_1  =  1.0
    r_3_2  =  1.0
    r_3_4  =  1.0
    r_3_5  =  1.0
    r_3_6  =  1.0
    r_3_8  =  1.0
    r_3_9  =  1.0
    r_4_1  =  1.0
    r_4_2  =  1.0
    r_4_5  =  1.0
    r_4_6  =  1.0
    r_4_8  =  1.0
    r_4_9  =  1.0
    r_5_1  =  1.0
    r_5_9  =  1.0
    r_6_1  =  1.0
    r_6_5  =  1.0
    r_6_9  =  1.0
    r_7_1  =  1.0
    r_7_2  =  1.0
    r_7_3  =  1.0
    r_7_4  =  1.0
    r_7_5  =  1.0
    r_7_6  =  1.0
    r_7_8  =  1.0
    r_7_9  =  1.0
    r_8_1  =  1.0
    r_8_5  =  1.0
    r_8_6  =  1.0
    r_8_9  =  1.0
    r_10_1  =  1.0
    r_10_2  =  1.0
    r_10_3  =  1.0
    r_10_4  =  1.0
    r_10_5  =  1.0
    r_10_6  =  1.0
    r_10_7  =  1.0
    r_10_8  =  1.0
    r_10_9  =  1.0
    --
    (x, y)  =  (3.0, 2.2500000000000027)
    (x, y)  =  (3.0, 4.25)
    (x, y)  =  (3.0, 5.249999999999999)
    (x, y)  =  (3.0, 4.75)
    (x, y)  =  (3.0, 2.7500000000000018)
    (x, y)  =  (3.0, 3.250000000000001)
    (x, y)  =  (3.0, 5.749999999999998)
    (x, y)  =  (3.0, 3.7500000000000004)
    (x, y)  =  (3.0, 1.000000000000003)
    (x, y)  =  (3.0, 6.999999999999998)
    (x, y)  =  (5.0, 3.9999999999999964)
    (x, y)  =  (1.0, 3.999999999999999)
    ```

* garment:

    - Width = 25
    - Height = 45.60
    - OPT: 8585.08312749297
    - Sol: 
    ```
    z_1_1  =  1.0
    z_2_3  =  1.0
    z_3_3  =  1.0
    z_4_1  =  1.0
    z_5_1  =  1.0
    z_6_1  =  1.0
    z_7_1  =  1.0
    z_8_3  =  1.0
    z_9_1  =  1.0
    z_10_1  =  1.0
    z_11_3  =  1.0
    --
    r_1_4  =  1.0
    r_1_5  =  1.0
    r_1_7  =  1.0
    r_1_9  =  1.0
    r_1_10  =  1.0
    r_2_11  =  1.0
    r_3_2  =  1.0
    r_3_9  =  1.0
    r_3_11  =  1.0
    r_4_5  =  1.0
    r_4_9  =  1.0
    r_4_10  =  1.0
    r_6_1  =  1.0
    r_6_4  =  1.0
    r_6_5  =  1.0
    r_6_7  =  1.0
    r_6_9  =  1.0
    r_6_10  =  1.0
    r_7_4  =  1.0
    r_7_5  =  1.0
    r_7_9  =  1.0
    r_7_10  =  1.0
    r_8_2  =  1.0
    r_8_3  =  1.0
    r_8_5  =  1.0
    r_8_10  =  1.0
    r_8_11  =  1.0
    r_9_5  =  1.0
    r_9_10  =  1.0
    r_10_5  =  1.0
    --
    (x, y)  =  (6.901206140350876, 32.70370358601185)
    (x, y)  =  (19.39473684210526, 17.680483118309084)
    (x, y)  =  (19.39473684210526, 30.142653229285127)
    (x, y)  =  (6.901206140350876, 16.170724034382495)
    (x, y)  =  (6.901206140350876, 2.434357075898886)
    (x, y)  =  (6.901206140350876, 40.057490586123066)
    (x, y)  =  (6.901206140350876, 25.34991658590063)
    (x, y)  =  (19.39473684210526, 39.73258303595937)
    (x, y)  =  (6.901206140350876, 10.617709210505408)
    (x, y)  =  (6.901206140350876, 6.785045838033652)
    (x, y)  =  (19.39473684210526, 4.470412924983333)
    ```

**Decision variables**:

$z_{ik} = \begin{cases}
1 & \text{ if department } i \text{ is assigned to bay } k\\
0 & \text{ otherwise}\\
\end{cases}$

$r_{ij} = \begin{cases}
1 & \text{ if department } i \text{ is above department } j \text{ in the same bay}\\
0 & \text{ otherwise}\\
\end{cases}$

$(x_i, y_i) = \text{ coordinates of the centroid of department } i$

In [66]:
for i in dpts:
    for k in max_bays:

        zs_val = uaflp_sol.get_var_value(zs[i, k])
        if zs_val == 1.0:
            print(f'z_{i}_{k}', ' = ', zs_val)

z_1_1  =  1.0
z_2_3  =  1.0
z_3_3  =  1.0
z_4_1  =  1.0
z_5_1  =  1.0
z_6_1  =  1.0
z_7_1  =  1.0
z_8_3  =  1.0
z_9_1  =  1.0
z_10_1  =  1.0
z_11_3  =  1.0


In [67]:
for i in dpts:
    for j in dpts:

        rs_val = uaflp_sol.get_var_value(rs[i, j])
        if rs_val == 1.0:
            print(f'r_{i}_{j}', ' = ', rs_val)

r_1_4  =  1.0
r_1_5  =  1.0
r_1_7  =  1.0
r_1_9  =  1.0
r_1_10  =  1.0
r_2_11  =  1.0
r_3_2  =  1.0
r_3_9  =  1.0
r_3_11  =  1.0
r_4_5  =  1.0
r_4_9  =  1.0
r_4_10  =  1.0
r_6_1  =  1.0
r_6_4  =  1.0
r_6_5  =  1.0
r_6_7  =  1.0
r_6_9  =  1.0
r_6_10  =  1.0
r_7_4  =  1.0
r_7_5  =  1.0
r_7_9  =  1.0
r_7_10  =  1.0
r_8_2  =  1.0
r_8_3  =  1.0
r_8_5  =  1.0
r_8_10  =  1.0
r_8_11  =  1.0
r_9_5  =  1.0
r_9_10  =  1.0
r_10_5  =  1.0


In [68]:
for i in dpts:

    x_coord = uaflp_sol.get_var_value(x[i])
    y_coord = uaflp_sol.get_var_value(y[i])
    print('(x, y)', ' = ', (x_coord, y_coord))

(x, y)  =  (6.901206140350876, 32.70370358601185)
(x, y)  =  (19.39473684210526, 17.680483118309084)
(x, y)  =  (19.39473684210526, 30.142653229285127)
(x, y)  =  (6.901206140350876, 16.170724034382495)
(x, y)  =  (6.901206140350876, 2.434357075898886)
(x, y)  =  (6.901206140350876, 40.057490586123066)
(x, y)  =  (6.901206140350876, 25.34991658590063)
(x, y)  =  (19.39473684210526, 39.73258303595937)
(x, y)  =  (6.901206140350876, 10.617709210505408)
(x, y)  =  (6.901206140350876, 6.785045838033652)
(x, y)  =  (19.39473684210526, 4.470412924983333)


### Best FSBP Model

MIP-FSBP model + constraints (23)-(27)

In [3]:
relaxed = False

In [4]:
# Original MIP-FBS Formulation
# Define UAFLP-MIP model
uaflp_mip = Model('UAFLP-MIP-KONAK-2006')

# Add variables
# Binary vars
if not relaxed:
    zs = {(i, k): uaflp_mip.binary_var(name=f'z_{i}_{k}')\
        for i in dpts for k in max_bays}

    rs = {(i, j): uaflp_mip.binary_var(name=f'r_{i}_{j}')\
        for i in dpts for j in dpts}

    ds = {(k): uaflp_mip.binary_var(name=f'd_{k}')\
        for k in max_bays}
else:
    zs = {(i, k): uaflp_mip.continuous_var(name=f'z_{i}_{k}', lb=0)\
      for i in dpts for k in max_bays}

    rs = {(i, j): uaflp_mip.continuous_var(name=f'r_{i}_{j}', lb=0)\
        for i in dpts for j in dpts}

    ds = {(k): uaflp_mip.continuous_var(name=f'd_{k}', lb=0)\
        for k in max_bays}

# Continuous vars
w = {(k): uaflp_mip.continuous_var(name=f'w_{k}', lb=0)\
      for k in max_bays}

ly = {(i): uaflp_mip.continuous_var(name=f'l_y_{i}', lb=0)\
      for i in dpts}

h = {(i, k): uaflp_mip.continuous_var(name=f'h_{i}_{k}', lb=0)\
       for i in dpts for k in max_bays}

x = {(i): uaflp_mip.continuous_var(name=f'x_{i}', lb=0)\
      for i in dpts}

y = {(i): uaflp_mip.continuous_var(name=f'y_{i}', lb=0)\
      for i in dpts}

dx = {(i, j): uaflp_mip.continuous_var(name=f'd_x_{i}_{j}', lb=0)\
        for i in dpts for j in dpts}

dy = {(i, j): uaflp_mip.continuous_var(name=f'd_y_{i}_{j}', lb=0)\
        for i in dpts for j in dpts}

# Add objective function
total_mf_costs = uaflp_mip.sum(uaflp_mip.sum(mat_flows[i-1, j-1] * (dx[i, j] + dy[i, j]) for j in dpts if i < j) for i in dpts)
uaflp_mip.minimize(total_mf_costs)

# Add constraints
for i in dpts:
    for j in dpts:
        if i < j:
            uaflp_mip.add_constraint(dx[i, j] >= x[i] - x[j])
            uaflp_mip.add_constraint(dx[i, j] >= x[j] - x[i])
            uaflp_mip.add_constraint(dy[i, j] >= y[i] - y[j])
            uaflp_mip.add_constraint(dy[i, j] >= y[j] - y[i])

for i in dpts:
    uaflp_mip.add_constraint(uaflp_mip.sum(zs[i, k] for k in max_bays) == 1)

for k in max_bays:
    uaflp_mip.add_constraint(w[k] == uaflp_mip.sum(zs[i, k] * dp_areas[i-1] for i in dpts)/height)

for i in dpts:
    for k in max_bays:
        uaflp_mip.add_constraint(min_side[i-1] * zs[i, k] <= w[k])
        uaflp_mip.add_constraint(w[k] <= max_side[i-1] + width * (1 - zs[i, k]))

for i in dpts:
    for k in max_bays:
        uaflp_mip.add_constraint(x[i] >= uaflp_mip.sum(w[j] for j in max_bays if j <= k) - 0.5*w[k] - \
                                 (width - min_side[i-1])*(1 - zs[i, k]))
        
        uaflp_mip.add_constraint(x[i] <= uaflp_mip.sum(w[j] for j in max_bays if j <= k) - 0.5*w[k] + \
                                 (width - min_side[i-1])*(1 - zs[i, k]))
        
for i in dpts:
    for j in dpts:
        if i < j:
            for k in max_bays:
                uaflp_mip.add_constraint(h[i, k]/dp_areas[i-1] - h[j, k]/dp_areas[j-1] - \
                                         np.max([max_side[i-1]/dp_areas[i-1], max_side[j-1]/dp_areas[j-1]])*(2 - zs[i, k] - zs[j, k]) <= 0)
                uaflp_mip.add_constraint(h[i, k]/dp_areas[i-1] - h[j, k]/dp_areas[j-1] + \
                                         np.max([max_side[i-1]/dp_areas[i-1], max_side[j-1]/dp_areas[j-1]])*(2 - zs[i, k] - zs[j, k]) >= 0)

for k in max_bays:
    uaflp_mip.add_constraint(uaflp_mip.sum(h[i, k] for i in dpts) == height*ds[k])

for i in dpts:
    for k in max_bays:
        uaflp_mip.add_constraint(min_side[i-1]*zs[i, k] <= h[i, k])
        uaflp_mip.add_constraint(h[i, k] <= max_side[i-1]*zs[i, k])

for i in dpts:
    uaflp_mip.add_constraint(uaflp_mip.sum(h[i, k] for k in max_bays) == ly[i])

for i in dpts:
    for j in dpts:
        if i != j:
            uaflp_mip.add_constraint(y[i] - 0.5*ly[i] >= y[j] + 0.5*ly[j] - height*(1 - rs[i, j]))

for i in dpts:
    for j in dpts:
        if i < j:
            uaflp_mip.add_constraint(rs[i, j] + rs[j, i] <= 1)

for i in dpts:
    for j in dpts:
        if i < j:
            for k in max_bays:
                uaflp_mip.add_constraint(rs[i, j] + rs[j, i] >= zs[i, k] + zs[j, k] - 1)

for i in dpts:
    uaflp_mip.add_constraint(0.5*ly[i] <= y[i])
    uaflp_mip.add_constraint(y[i] <= height - 0.5*ly[i])

In [8]:
uaflp_mip.pprint()

// This file has been generated by DOcplex
// model name is: UAFLP-MIP-KONAK-2006
// single vars section
dvar bool z_1_1;
dvar bool z_1_2;
dvar bool z_1_3;
dvar bool z_1_4;
dvar bool z_1_5;
dvar bool z_1_6;
dvar bool z_1_7;
dvar bool z_1_8;
dvar bool z_1_9;
dvar bool z_2_1;
dvar bool z_2_2;
dvar bool z_2_3;
dvar bool z_2_4;
dvar bool z_2_5;
dvar bool z_2_6;
dvar bool z_2_7;
dvar bool z_2_8;
dvar bool z_2_9;
dvar bool z_3_1;
dvar bool z_3_2;
dvar bool z_3_3;
dvar bool z_3_4;
dvar bool z_3_5;
dvar bool z_3_6;
dvar bool z_3_7;
dvar bool z_3_8;
dvar bool z_3_9;
dvar bool z_4_1;
dvar bool z_4_2;
dvar bool z_4_3;
dvar bool z_4_4;
dvar bool z_4_5;
dvar bool z_4_6;
dvar bool z_4_7;
dvar bool z_4_8;
dvar bool z_4_9;
dvar bool z_5_1;
dvar bool z_5_2;
dvar bool z_5_3;
dvar bool z_5_4;
dvar bool z_5_5;
dvar bool z_5_6;
dvar bool z_5_7;
dvar bool z_5_8;
dvar bool z_5_9;
dvar bool z_6_1;
dvar bool z_6_2;
dvar bool z_6_3;
dvar bool z_6_4;
dvar bool z_6_5;
dvar bool z_6_6;
dvar bool z_6_7;
dvar bool z

In [ ]:
# Solve original model
uaflp_sol = uaflp_mip.solve(clean_before_solve=True, log_output=True)

In [ ]:
# Show solution for the original model
print(uaflp_sol.solve_status)
print(uaflp_mip.solve_details.time)
print(uaflp_mip.solve_details.nb_nodes_processed)
print(uaflp_sol.get_objective_value())

In [ ]:
for i in dpts:
    for k in max_bays:

        zs_val = uaflp_sol.get_var_value(zs[i, k])
        if zs_val > 0:
            print(f'z_{i}_{k}', ' = ', zs_val)

In [ ]:
for i in dpts:
    for j in dpts:

        dx_val = uaflp_sol.get_var_value(dx[i, j])
        dy_val = uaflp_sol.get_var_value(dy[i, j])
        if dx_val != 0 or dy_val != 0:
            print(dx_val, dy_val)

#### Adding improvements to the model

In [ ]:
# Tightening LP lower bound
# for i in dpts:
#     for j in dpts:
#         if i < j:
#             for k in max_bays:
#                 for m in max_bays:
#                     if k != m:
#                         uaflp_mip.add_constraint(dx[i, j] >= 0.5*(min_side[i-1] + min_side[j-1]) + width*(zs[i, k] + zs[j, m] - 2))

for i in dpts:
    for j in dpts:
        if i < j:
            for k in max_bays:
                uaflp_mip.add_constraint(dy[i, j] >= 0.5*(min_side[i-1] + min_side[j-1]) + height*(zs[i, k] + zs[j, k] - 2))

In [5]:
# Reducing model degeneracy: fill bays sequentially
for k in max_bays:
    if k < len(max_bays):
        uaflp_mip.add_constraint(n_dpts*uaflp_mip.sum(zs[i, k] for i in dpts) >= uaflp_mip.sum(zs[i, k+1] for i in dpts))

In [ ]:
# Tightened dx constraint
for i in dpts:
    for j in dpts:
        if i < j:
            for k in max_bays:
                for m in max_bays:
                    if k != m:
                        uaflp_mip.add_constraint(dx[i, j] >= 0.5*(min_side[i-1] + min_side[j-1]) + \
                                                 uaflp_mip.min(min_side[p-1] for p in dpts if p != i and p != j)*(np.abs(m - k) - 1) + \
                                                    len(max_bays)*width*(zs[i, k] + zs[j, m] - 2))

In [6]:
# Solve original model + improvements
uaflp_sol = uaflp_mip.solve(clean_before_solve=True, log_output=True)

Version identifier: 22.1.0.0 | 2022-03-09 | 1a383f8ce
CPXPARAM_Read_DataCheck                          1
Tried aggregator 1 time.
MIP Presolve eliminated 53 rows and 85 columns.
MIP Presolve modified 1184 coefficients.
Reduced MIP has 1093 rows, 227 columns, and 4047 nonzeros.
Reduced MIP has 114 binaries, 0 generals, 0 SOSs, and 0 indicators.
Presolve time = 0.02 sec. (2.96 ticks)
Probing fixed 2 vars, tightened 4 bounds.
Probing time = 0.02 sec. (1.98 ticks)
Cover probing fixed 0 vars, tightened 15 bounds.
Tried aggregator 1 time.
Detecting symmetries...
MIP Presolve eliminated 0 rows and 2 columns.
MIP Presolve modified 6 coefficients.
Reduced MIP has 1093 rows, 225 columns, and 4045 nonzeros.
Reduced MIP has 112 binaries, 0 generals, 0 SOSs, and 0 indicators.
Presolve time = 0.00 sec. (2.17 ticks)
Probing time = 0.00 sec. (1.78 ticks)
Cover probing fixed 0 vars, tightened 36 bounds.
Clique table members: 245.
MIP emphasis: balance optimality and feasibility.
MIP search method: dyna

In [7]:
# Show solution for the original model
print(uaflp_sol.solve_status)
print(uaflp_mip.solve_details.time)
print(uaflp_mip.solve_details.nb_nodes_processed)
print(uaflp_sol.get_objective_value())

JobSolveStatus.OPTIMAL_SOLUTION
1.0
1495
138.89010989010987


In [9]:
for i in dpts:
    for k in max_bays:

        zs_val = uaflp_sol.get_var_value(zs[i, k])
        if zs_val > 0:
            print(f'z_{i}_{k}', ' = ', zs_val)

z_1_1  =  1.0
z_2_1  =  1.0
z_3_1  =  1.0
z_4_1  =  1.0
z_5_2  =  1.0
z_6_2  =  1.0
z_7_2  =  1.0


In [5]:
# Add a solution
uaflp_mip.add_constraint(zs[1, 1] == 1)
uaflp_mip.add_constraint(zs[2, 1] == 1)
uaflp_mip.add_constraint(zs[3, 1] == 1)
uaflp_mip.add_constraint(zs[4, 2] == 1)
uaflp_mip.add_constraint(zs[5, 2] == 1)
uaflp_mip.add_constraint(zs[6, 2] == 1)
uaflp_mip.add_constraint(zs[7, 1] == 1)

uaflp_mip.add_constraint(rs[1, 2] == 1)
uaflp_mip.add_constraint(rs[1, 3] == 1)
uaflp_mip.add_constraint(rs[1, 7] == 1)
uaflp_mip.add_constraint(rs[2, 3] == 1)
uaflp_mip.add_constraint(rs[2, 7] == 1)
uaflp_mip.add_constraint(rs[3, 7] == 1)
uaflp_mip.add_constraint(rs[4, 5] == 1)
uaflp_mip.add_constraint(rs[4, 6] == 1)
uaflp_mip.add_constraint(rs[5, 6] == 1)

uaflp_mip.add_constraint(ds[1] == 1)
uaflp_mip.add_constraint(ds[2] == 1)

docplex.mp.LinearConstraint[](d_2,EQ,1)

In [9]:
for i in dpts:
    for j in dpts:
        if i != j:
            dx_val = uaflp_sol.get_var_value(dx[i, j])
            dy_val = uaflp_sol.get_var_value(dy[i, j])
            if dx_val != 0 or dy_val != 0:
                print(f'dpts {i}-{j}: ')
                print(dx_val, dy_val)
                print('--')

dpts 1-2: 
0 3.6491228070175428
--
dpts 1-3: 
0 7.298245614035087
--
dpts 1-4: 
4.269230769230769 2.50877192982456
--
dpts 1-5: 
4.269230769230769 7.925438596491226
--
dpts 1-6: 
4.269230769230769 10.092105263157894
--
dpts 1-7: 
0 10.149122807017543
--
dpts 2-3: 
0 3.6491228070175445
--
dpts 2-4: 
4.269230769230769 1.140350877192983
--
dpts 2-5: 
4.269230769230769 4.276315789473683
--
dpts 2-6: 
4.269230769230769 6.442982456140352
--
dpts 2-7: 
0 6.5
--
dpts 3-4: 
4.269230769230769 4.789473684210527
--
dpts 3-5: 
4.269230769230769 0.6271929824561386
--
dpts 3-6: 
4.269230769230769 2.7938596491228074
--
dpts 3-7: 
0 2.8508771929824555
--
dpts 4-5: 
0 5.416666666666666
--
dpts 4-6: 
0 7.583333333333335
--
dpts 4-7: 
4.269230769230769 7.640350877192983
--
dpts 5-6: 
0 2.1666666666666687
--
dpts 5-7: 
4.269230769230769 2.223684210526317
--
dpts 6-7: 
4.269230769230769 0.0570175438596483
--


In [10]:
for i in dpts:
    
    x_val = uaflp_sol.get_var_value(x[i])
    y_val = uaflp_sol.get_var_value(y[i])
    print(f'dpt {i}: ')
    print(x_val, y_val)
    print('--')

dpt 1: 
2.1923076923076925 11.175438596491228
--
dpt 2: 
2.1923076923076925 7.526315789473685
--
dpt 3: 
2.1923076923076925 3.8771929824561404
--
dpt 4: 
6.461538461538462 8.666666666666668
--
dpt 5: 
6.461538461538462 3.2500000000000018
--
dpt 6: 
6.461538461538462 1.0833333333333333
--
dpt 7: 
2.1923076923076925 1.026315789473685
--
